### PR TITLE
fix: timestamps:false on bulkWrite works

### LINKED ISSUE #13611: timestamps options does not work as expect in subdocuments when using Model.bulkWrite to do updateMany operator.
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

7.3.4

### Node.js version

v14.21.3

### MongoDB server version

6.0.6

### Typescript version (if applicable)

5.1.6

### Description

Following the documentation [Disabling Timestamps](https://mongoosejs.com/docs/timestamps.html#disabling-timestamps)

There are two issues:

1. Runtime issue. Try to disable updating the timestamps on the `post` subdocuments. But the `updatedAt` is still updated after setting the `timestamps` option to `false`.

2. Compile time issue. When setting the `timestamps` option, TSC throws an error:

```txt
Type '{ filter: { isDeleted: false; }; update: { 'posts.$[post].isDeleted': true; }; arrayFilters: { 'post.category': { $eq: string; }; }[]; upsert: false; timestamps: false; }' is not assignable to type 'UpdateManyModel<any>'.
  Object literal may only specify known properties, and 'timestamps' does not exist in type 'UpdateManyModel<any>'.ts(2322)
```

When I checkout the type, there is no `timestamps` property on the `UpdateManyModel` interface.

```ts
export declare interface UpdateManyModel<TSchema extends Document = Document> {
    /** The filter to limit the updated documents. */
    filter: Filter<TSchema>;
    /** A document or pipeline containing update operators. */
    update: UpdateFilter<TSchema> | UpdateFilter<TSchema>[];
    /** A set of filters specifying to which array elements an update should apply. */
    arrayFilters?: Document[];
    /** Specifies a collation. */
    collation?: CollationOptions;
    /** The index to use. If specified, then the query system will only consider plans using the hinted index. */
    hint?: Hint;
    /** When true, creates a new document if no document matches the query. */
    upsert?: boolean;
}
```

### Steps to Reproduce

```ts
import mongoose from 'mongoose';
import { config } from '../../config';

mongoose.set('debug', true);
console.log(mongoose.version);

const PostSchema = new mongoose.Schema(
	{
		category: { type: String, required: true },
		title: { type: String, required: true },
		isDeleted: { type: Boolean, default: false },
	},
	{ timestamps: true },
);
const Post = mongoose.model('post', PostSchema);

const UserSchema = new mongoose.Schema(
	{
		username: { type: String, required: true },
		posts: { type: [PostSchema] },
		isDeleted: { type: Boolean, default: false },
	},
	{ timestamps: true },
);
const User = mongoose.model('user', UserSchema);

(async function main() {
	try {
		await mongoose.connect(config.MONGODB_URI);
		// seed
		const [u1] = await User.create([
			{
				username: 'user a',
				posts: [
					{ title: 'post a', category: 'a', isDeleted: false },
					{ title: 'post b', category: 'b', isDeleted: false },
				],
				isDeleted: false,
			},
		]);

		console.log(u1?.posts[0].updatedAt);

		// test
		await User.bulkWrite([
			{
				updateMany: {
					filter: {
						isDeleted: false,
					},
					update: {
						'posts.$[post].isDeleted': true,
					},
					arrayFilters: [
						{
							'post.category': { $eq: 'a' },
						},
					],
					upsert: false,
					timestamps: false,
				},
			},
		]);
		const u2 = await User.findOne({ _id: u1._id });
		console.log(u2?.posts[0].updatedAt);
	} catch (error) {
		console.error(error);
	} finally {
		await mongoose.connection.close();
	}
})();
```

Logs:
```zsh
Mongoose: users.insertOne({ username: 'user a', posts: [ { category: 'a', title: 'post a', isDeleted: false, _id: ObjectId("64b11d543062f023ceebd09e"), createdAt: new Date("Fri, 14 Jul 2023 10:03:00 GMT"), updatedAt: new Date("Fri, 14 Jul 2023 10:03:00 GMT") }, { category: 'b', title: 'post b', isDeleted: false, _id: ObjectId("64b11d543062f023ceebd09f"), createdAt: new Date("Fri, 14 Jul 2023 10:03:00 GMT"), updatedAt: new Date("Fri, 14 Jul 2023 10:03:00 GMT") } ], isDeleted: false, _id: ObjectId("64b11d543062f023ceebd09d"), createdAt: new Date("Fri, 14 Jul 2023 10:03:00 GMT"), updatedAt: new Date("Fri, 14 Jul 2023 10:03:00 GMT"), __v: 0}, {})
2023-07-14T10:03:00.285Z
Mongoose: users.bulkWrite([ { updateMany: { filter: { isDeleted: false }, update: { '$set': { 'posts.$[post].updatedAt': 2023-07-14T10:03:00.857Z, 'posts.$[post].isDeleted': true } }, arrayFilters: [ { 'post.category': { '$eq': 'a' } } ], upsert: false, timestamps: false } }], {})
Mongoose: users.findOne({ _id: ObjectId("64b11d543062f023ceebd09d") }, {})
2023-07-14T10:03:00.857Z
```

As you can see, `2023-07-14T10:03:00.285Z` has been changed to  `2023-07-14T10:03:00.857Z`

### Expected Behavior

The `updatedAt` should stay the same after updating.

### PR DESCRIPTION
**Pull request by IslandRhythms**
closes #13611

### CODE REVIEW

File: test/model.test.js [hunk 1]
```diff
@@ -5732,6 +5732,51 @@ describe('Model', function() {
...
+    const User = db.model('gh13611User', userSchema);
+
+    const entry = await User.create({
+      name: 'Test Testerson',
+      posts: [{ title: 'title a', category: 'a', isDeleted: false }, { title: 'title b', category: 'b', isDeleted: false }],
+      isDeleted: false
+    });
+    const initialTime = entry.posts[0].updatedAt;
+
+    await User.bulkWrite([{
```
Comment (AbdelrahmanHafez) [inline test/model.test.js:5758]: Can we sleep for 5 milliseconds here to guarantee that the `bulkWrite` wouldn't be executed at the same exact millisecond and prevent false-positives in the future?

File: lib/helpers/model/castBulkWrite.js [hunk 2]
```diff
@@ -66,7 +65,9 @@ module.exports = function castBulkWrite(originalModel, op, options) {
           applyTimestampsToUpdate(now, createdAt, updatedAt, op['updateOne']['update'], {});
         }
 
-        applyTimestampsToChildren(now, op['updateOne']['update'], model.schema);
+        if (op['updateOne'].timestamps || op['updateOne'].timestamps == null) {
```
Comment (AbdelrahmanHafez) [inline lib/helpers/model/castBulkWrite.js:68]: Nitpick:
```suggestion
        if (op['updateOne'].timestamps === true || op['updateOne'].timestamps == null) {
```

File: lib/helpers/model/castBulkWrite.js [hunk 3]
```diff
@@ -117,8 +118,9 @@ module.exports = function castBulkWrite(originalModel, op, options) {
           const updatedAt = model.schema.$timestamps.updatedAt;
           applyTimestampsToUpdate(now, createdAt, updatedAt, op['updateMany']['update'], {});
         }
-
-        applyTimestampsToChildren(now, op['updateMany']['update'], model.schema);
+        if (op['updateMany'].timestamps || op['updateMany'].timestamps == null) {
```
Comment (AbdelrahmanHafez) [inline lib/helpers/model/castBulkWrite.js:121]: Nitpick:
```suggestion
        if (op['updateMany'].timestamps === true || op['updateMany'].timestamps == null) {
```

File: test/model.test.js (hunk 1 above)
Comment (vkarpov15) [reply test/model.test.js:5758]: :+1: for this

File: lib/helpers/model/castBulkWrite.js (hunk 3 above)
Comment (vkarpov15) [reply lib/helpers/model/castBulkWrite.js:121]: Can we do `if (op['updateMany'].timestamps !== false)` to be consistent with the other `if` statements in this file?

File: lib/helpers/model/castBulkWrite.js (hunk 2 above)
Comment (vkarpov15) [reply lib/helpers/model/castBulkWrite.js:68]: Can we do `if (op['updateMany'].timestamps !== false)` to be consistent with the other `if` statements in this file?

**Review by vkarpov15 — CHANGES_REQUESTED** [review]

File: test/model.test.js [hunk 4]
```diff
@@ -5732,6 +5732,52 @@ describe('Model', function() {
...
+
+    const User = db.model('gh13611User', userSchema);
+
+    const entry = await User.create({
+      name: 'Test Testerson',
+      posts: [{ title: 'title a', category: 'a', isDeleted: false }, { title: 'title b', category: 'b', isDeleted: false }],
+      isDeleted: false
+    });
+    const initialTime = entry.posts[0].updatedAt;
+    await delay(5000);
```
Comment (AbdelrahmanHafez) [inline test/model.test.js:5757]: 5 seconds is a bit of an overkill. Just 1 millisecond should be enough to guarantee, I'd go for an arbitrary 5 milliseconds.

Comment (IslandRhythms) [reply test/model.test.js:5757]: I totally misread your initial comment

**Review by vkarpov15 — APPROVED** [review]

**Pull request merged**